src/hotspot/share/runtime/os.cpp

```diff
@@ -1857,3 +1845,73 @@ void os::naked_sleep(jlong millis) {
   }
   naked_short_sleep(millis);
 }
+
+
+////// Implementation of PagesizeSet
+
+void os::PagesizeSet::add(size_t pagesize) {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  _v |= pagesize;
+}
+
+bool os::PagesizeSet::is_set(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  return (_v & pagesize) > 0;
+}
+
+size_t os::PagesizeSet::next_smaller(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  // mask out all pages sizes >= pagesize:
+  uintx v2 = _v & (pagesize - 1);
+  if (v2 > 0) {
+    return round_down_power_of_2(v2);
+  }
+  return 0;
```

Comment:
In next_larger you use the for:
```
if (v2 == 0) {
  return 0;
}
return ...
```
I think it would be nice to be consistent between the two functions.

Reply:
> Just a few drive-by comment: Could PagesizeSet be named PageSizeSet. The rest of the code separates the word, so I think that name would be better. The same for some of the parameters that are named pagesize.
> 
> Also, I think you could have called the class PageSizes. I don't think the fact that this is a 'set' matters for the usages of the class. It also makes the variable name match the type name, which I guess is an indication that it wouldn't be such a bad name:
> 
> ```
> static PageSizes        _page_sizes;`
> ```
> 
> vs
> 
> ```
>   static PagesizeSet        _page_sizes;
> ```

Hi Stefan,

thanks for your review! I adapted most of your suggestions.

Cheers, Thomas